src/lib.rs

```diff
@@ -1,7 +1,334 @@
+use hashbrown::HashMap;
+use hint::unreachable_unchecked;
+use rand::{RngCore, SeedableRng};
+use rand_xoshiro::Xoshiro128PlusPlus;
 
 #[cfg(feature = "binary-vendor")]
 include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
+
+const MAX_NUMBER_OF_TXS: usize = 2usize.pow(16);
+
+static mut STATE: Option<Contract> = None;
+
+#[derive(Default, Debug)]
+struct Contract {
+    admin: ActorId,
+
+    fungible_token: Option<ActorId>,
+    started: u64,
+    ending: u64,
+    players: Vec<ActorId>,
+    prize_fund: u128,
+    participation_cost: u128,
+    is_active: bool,
+
+    winner: Option<ActorId>,
+
+    txs_for_actor: BTreeMap<u64, ActorId>,
+    actors_for_tx: HashMap<ActorId, u64>,
+    tx_id_nonce: u64,
+}
+
+impl Contract {
+    fn start(
+        &mut self,
+        duration: u64,
+        participation_cost: u128,
+        fungible_token: Option<ActorId>,
+    ) -> Result<Event, Error> {
+        if self.admin != msg::source() {
+            return Err(Error::AccessRestricted);
+        }
+
+        if self.is_active {
+            return Err(Error::UnexpectedGameStatus);
+        }
+
+        if matches!(fungible_token, Some(fungible_token) if fungible_token.is_zero()) {
+            return Err(Error::ZeroActorId);
+        }
+
+        self.players.clear();
+
+        self.winner = None;
+        self.prize_fund = 0;
+        self.started = exec::block_timestamp();
+        self.ending = self.started.saturating_add(duration);
+        self.participation_cost = participation_cost;
+        self.fungible_token = fungible_token;
+        self.is_active = true;
+
+        // TODO: uncomment and update doc & tests after closing
+        // https://github.com/gear-tech/gear/issues/1781.
+        // msg::send_delayed(
+        //     exec::program_id(),
+        //     Action::PickWinner,
+        //     0,
+        //     (duration / 1000) as u32,
+        // )?;
+
+        Ok(Event::Started {
+            ending: self.ending,
+            participation_cost,
+            fungible_token,
+        })
+    }
+
+    async fn pick_winner(&mut self) -> Result<Event, Error> {
+        if !self.is_active {
+            return Err(Error::UnexpectedGameStatus);
+        }
+
+        let msg_source = msg::source();
+        let exec_program = exec::program_id();
+        let block_timestamp = exec::block_timestamp();
+
+        if msg_source == self.admin {
+            if self.ending > block_timestamp {
+                return Err(Error::UnexpectedGameStatus);
+            }
+        } else if msg_source != exec_program {
+            return Err(Error::AccessRestricted);
+        }
+
+        let winner = self.winner.unwrap_or_else(|| {
+            let winner = if self.players.is_empty() {
+                ActorId::zero()
+            } else {
+                let mut random_data = [0; (usize::BITS / 8) as usize];
+
+                Xoshiro128PlusPlus::seed_from_u64(block_timestamp).fill_bytes(&mut random_data);
```

Comment:
Can we use [`exec::random`](https://docs.gear.rs/gstd/exec/fn.random.html) here?

Reply:
Yes, but I'd like to stick with the current implementation because it makes the winner prediction easier in test environment.